src/goto-instrument/unwind.cpp

```diff
@@ -45,136 +163,254 @@ Function: unwind
     {
-      // guard against 'fall-out'
       goto_programt::targett t_goto=goto_program.insert_before(loop_exit);
-  
-      t_goto->make_goto(loop_exit);
+
+      t_goto->make_goto(get_mutable(goto_program, loop_exit));
       t_goto->source_location=loop_exit->source_location;
       t_goto->function=loop_exit->function;
       t_goto->guard=true_exprt();
     }
-  }
-  
-  goto_programt::targett t_skip=goto_program.insert_before(loop_exit);
-  goto_programt::targett loop_iter=t_skip;
-  
-  t_skip->make_skip();
-  t_skip->source_location=loop_head->source_location;
-  t_skip->function=loop_head->function;
-  
-  // record the exit point of first iteration
-  iteration_points[0]=loop_iter;
-  
-  // build a map for branch targets inside the loop
-  std::map<goto_programt::targett, unsigned> target_map;
 
-  { 
-    unsigned count=0;
-    for(goto_programt::targett t=loop_head;
-        t!=loop_exit; t++, count++)
+    // add a skip before the loop exit
+
+    goto_programt::targett t_skip=goto_program.insert_before(loop_exit);
+    t_skip->make_skip();
+    t_skip->source_location=loop_head->source_location;
+    t_skip->function=loop_head->function;
+
+    // where to go for the next iteration
+    goto_programt::targett loop_iter=t_skip;
+
+    // record the exit point of first iteration
+    iteration_points[0]=loop_iter;
+
+    // re-direct any branches that go to loop_head to loop_iter
+
+    for(goto_programt::targett t=get_mutable(goto_program, loop_head);
+        t!=loop_iter; t++)
     {
-      assert(t!=goto_program.instructions.end());
-      target_map[t]=count;
+      if(!t->is_goto())
+        continue;
+
+      if(t->get_target()==loop_head)
+        t->set_target(loop_iter);
     }
-  }
 
-  // re-direct any branches that go to loop_head to loop_iter
-  
-  for(goto_programt::targett t=loop_head;
-      t!=loop_iter; t++)
-  {
-    assert(t!=goto_program.instructions.end());
-    for(goto_programt::instructiont::targetst::iterator
-        t_it=t->targets.begin();
-        t_it!=t->targets.end();
-        t_it++)
-      if(*t_it==loop_head) *t_it=loop_iter;
-  }
-  
-  // we make k-1 copies, to be inserted before loop_exit
-  goto_programt copies;
+    // k-1 additional copies
+    for(unsigned i=1; i<k; i++)
+    {
+      goto_programt tmp_program;
+      copy_segment(loop_head, loop_exit, tmp_program);
+      assert(!tmp_program.instructions.empty());
 
-  for(unsigned i=1; i<k; i++)
-  {
-    // make a copy
-    std::vector<goto_programt::targett> target_vector;
-    target_vector.reserve(target_map.size());
+      iteration_points[i]=--tmp_program.instructions.end();
 
-    for(goto_programt::targett t=loop_head;
-        t!=loop_exit; t++)
-    {
-      assert(t!=goto_program.instructions.end());
-      goto_programt::targett copied_t=copies.add_instruction();
-      *copied_t=*t;
-      target_vector.push_back(copied_t);
+      copies.destructive_append(tmp_program);
     }
+  }
+  else
+  {
+    assert(k==0);
```

Comment:
If I read the code correctly, this is the else case of k!=0 - should be trivially true.